### PR TITLE
statistics: multidimensional hot peer cache

### PR DESCRIPTION
**Pull request by Luffbee**
### What problem does this PR solve? 

First step of #2139 . The hot peer cache is not working with multidimensional load.

### What is changed and how it works?

Modify the `TopN` to support multidimensional load.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: server/statistics/topn.go [hunk 1]
```diff
@@ -24,27 +24,30 @@ import (
...
+// NewTopN returns a k-dimensional TopN with given TTL.
+func NewTopN(k, n int, ttl time.Duration) *TopN {
+	ret := &TopN{
 		n:      maxInt(n, 1),
-		topn:   newTopNHeap(n),
-		rest:   newRevTopNHeap(n),
+		topns:  make([]*singleTopN, k),
 		ttlLst: newTTLList(ttl),
 	}
+	k = maxInt(k, 1)
```
Comment (rleungx) [inline server/statistics/topn.go]: Why do we need it?

File: server/statistics/hot_peer_cache.go [hunk 2]
```diff
@@ -28,12 +28,22 @@ const (
 
 	rollingWindowsSize = 5
 
-	hotWriteRegionMinBytesRate = 1 * 1024
-	hotReadRegionMinBytesRate  = 8 * 1024
-
 	hotRegionReportMinInterval = 3
 
-	hotRegionAntiCount = 1
+	hotRegionAntiCount = 2
```
Comment (rleungx) [inline server/statistics/hot_peer_cache.go]: Is it changed on purpose?

File: server/statistics/topn.go [hunk 3]
```diff
@@ -24,27 +24,30 @@ import (
...
-	rest   *indexedHeap
+	topns  []*singleTopN
 	ttlLst *ttlList
 }
 
-// NewTopN returns a TopN with given TTL.
-func NewTopN(n int, ttl time.Duration) *TopN {
-	return &TopN{
+// NewTopN returns a k-dimensional TopN with given TTL.
+func NewTopN(k, n int, ttl time.Duration) *TopN {
```
Comment (rleungx) [inline server/statistics/topn.go]: Do we need to support different sizes for the different dimensions?

File: server/statistics/topn.go (hunk 1 above)
Comment (Luffbee) [reply server/statistics/topn.go]: To make sure `k >= 1` after this line.

Comment (rleungx) [reply server/statistics/topn.go]: But if `k == 0`, it will cause panic?

Comment (Luffbee) [reply server/statistics/topn.go]: I see, it is bug.

Comment (Luffbee) [reply server/statistics/topn.go]: Fixed. I decided to make it panic if `k <= 0 || n <= 0`.

File: server/statistics/hot_peer_cache.go (hunk 2 above)
Comment (Luffbee) [reply server/statistics/hot_peer_cache.go]: Yes, see line 291, I change the `<` to `<=`.

File: server/statistics/topn.go (hunk 3 above)
Comment (Luffbee) [reply server/statistics/topn.go]: I don't think we need this.

File: server/statistics/topn.go [hunk 4]
```diff
@@ -54,61 +58,46 @@ func (tn *TopN) Len() int {
...
+	return tn.topns[k].GetAllTopN()
 }
 
 // GetAll returns all items.
 func (tn *TopN) GetAll() []TopNItem {
 	tn.rw.RLock()
 	defer tn.rw.RUnlock()
-	topn := tn.topn.GetAll()
-	return append(topn, tn.rest.GetAll()...)
+	return tn.topns[0].GetAll()
```
Comment (nolouch) [inline server/statistics/topn.go]: No need `GetAll(k int)`?

Comment (Luffbee) [reply server/statistics/topn.go]: No, all items in `TopN` will be in all dimensions.

File: server/statistics/hot_peer.go [hunk 5]
```diff
@@ -47,9 +55,16 @@ func (stat *HotPeerStat) ID() uint64 {
 }
 
 // Less compares two HotPeerStat.Implementing TopNItem.
-func (stat *HotPeerStat) Less(than TopNItem) bool {
+func (stat *HotPeerStat) Less(k int, than TopNItem) bool {
 	rhs := than.(*HotPeerStat)
-	return stat.BytesRate < rhs.BytesRate
+	switch k {
+	case keyDim:
+		return stat.GetKeysRate() < rhs.GetKeysRate()
```
[the remark this answers is not in the document]
Comment (Luffbee) [reply server/statistics/hot_peer.go]: Renamed.

File: server/statistics/topn.go (hunk 4 above)
Comment (nolouch) [reply server/statistics/topn.go]: Why only `topns[0]` in here?

Comment (Luffbee) [reply server/statistics/topn.go]: Any dimension contains all items, so `tn.topns[0].GetAll() == tn.topns[i].GetAll()` for `i = 1, 2, 3, ...`.

**Comment by Luffbee** [comment]
PTAL @rleungx

File: server/statistics/topn_test.go [hunk 6]
```diff
@@ -25,118 +26,238 @@ var _ = Suite(&testTopNSuite{})
...
+	// check Get
 	for i := uint64(0); i < Total; i++ {
 		it := tn.Get(i).(*item)
 		c.Assert(it.id, Equals, i)
-		c.Assert(it.value, Equals, -float64(i))
+		c.Assert(it.values[0], Equals, -float64(i))
+	}
+}
+
+func putPerm(c *C, tn *TopN, K, Total int, f func(x int) float64, isUpdate bool) {
```
Comment (rleungx) [inline server/statistics/topn_test.go]: `K, Total` should start with the lowercase.

File: server/statistics/topn.go [hunk 7]
```diff
@@ -24,27 +24,31 @@ import (
...
-func NewTopN(n int, ttl time.Duration) *TopN {
-	return &TopN{
-		n:      maxInt(n, 1),
-		topn:   newTopNHeap(n),
-		rest:   newRevTopNHeap(n),
+// NewTopN returns a k-dimensional TopN with given TTL.
+// NOTE: panic if k <= 0 or n <= 0.
+func NewTopN(k, n int, ttl time.Duration) *TopN {
+	if k <= 0 || n <= 0 {
+		panic("invalid arguments for NewTopN")
```
Comment (rleungx) [inline server/statistics/topn.go]: How about printing the value of `n` and `k`?

File: server/statistics/topn.go [hunk 8]
```diff
@@ -117,59 +106,129 @@ func (tn *TopN) RemoveExpired() {
...
 	tn.rw.Lock()
 	defer tn.rw.Unlock()
-	item := tn.removeItemLocked(id)
+	for _, stn := range tn.topns {
+		item = stn.Remove(id)
+	}
 	_ = tn.ttlLst.Remove(id)
 	tn.maintain()
-	return item
+	return
```
Comment (rleungx) [inline server/statistics/topn.go]: It seems no need to return

Comment (Luffbee) [reply server/statistics/topn.go]: I think return the removed item is a better API. And it is used in the test.

File: server/statistics/topn.go (hunk 7 above)
Comment (Luffbee) [reply server/statistics/topn.go]: Fixed.

File: server/statistics/topn_test.go (hunk 6 above)
Comment (Luffbee) [reply server/statistics/topn_test.go]: Renamed.

**Pull request merged**